api/agent/drivers/docker/docker.go

```diff
@@ -186,6 +202,80 @@ func containerCleaner(ctx context.Context, driver *DockerDriver) {
 	}
 }
 
+// syncImageCleaner lists the current images on the system and adds them to the
+// image cache. The operation is performed once during startup to ensure a
+// restart of the fn-agent keeps track of previous state.
+func syncImageCleaner(ctx context.Context, driver *DockerDriver) {
+	if driver.imgCache == nil {
+		return
+	}
+
+	const imageListTimeout = time.Duration(60 * time.Second)
+
+	ctx, log := common.LoggerWithFields(ctx, logrus.Fields{"stack": "syncImageCleaner"})
+	limiter := rate.NewLimiter(0.5, 1)
+
+	for limiter.Wait(ctx) == nil {
+		ctx, cancel := context.WithTimeout(ctx, imageListTimeout)
+		images, err := driver.docker.ListImages(docker.ListImagesOptions{Context: ctx})
+		cancel()
+
+		if err == nil {
+			for _, img := range images {
+				log.WithError(err).Errorf("ListImages add %+v", img)
+				driver.imgCache.Update(&CachedImage{
+					ID:       img.ID,
+					ParentID: img.ParentID,
+					RepoTags: img.RepoTags,
+					Size:     uint64(img.Size),
+				})
+			}
+			return
+		}
+
+		log.WithError(err).Error("ListImages error, will retry...")
+	}
+}
+
+// runImageCleaner runs continuously and monitors image cache state. If the
+// cache is over the high water mark limit, then it tries to remove least recently
+// used image.
+func runImageCleaner(ctx context.Context, driver *DockerDriver) {
+	if driver.imgCache == nil {
+		return
+	}
+
+	const removeImgTimeout = time.Duration(60 * time.Second)
+
+	ctx, log := common.LoggerWithFields(ctx, logrus.Fields{"stack": "runImageCleaner"})
+	limiter := rate.NewLimiter(0.5, 1)
+	notifier := driver.imgCache.GetNotifier()
+
+	for limiter.Wait(ctx) == nil {
```

Comment:
limiter tries to slow down image removals. To reduce too many docker calls executing. I can increase the rate a little bit (I think 0.5 is too slow.)

Reply:
Increased these to 2 ops/sec now. It's just a safety rate number in case a large set of images are to be removed and we encounter errors that put them back to LRU, etc.